### ISSUE TITLE
False positive for `boxed_local` with trait default methods

### ISSUE THREAD

**Issue by dmarcuse** [rust-lang/rust-clippy#3739]
When defining a trait that has a method accepting a `self: Box<Self>` which has a default implementation on the trait itself, the `boxed_local` lint is triggered. This was fixed for impls in #3321, but still triggers for traits.

Clippy version: `clippy 0.0.212 (b2601be 2018-11-27)`

<details>
<summary>Example Code</summary>

```rust
trait ExampleTrait {
    fn method(self: Box<Self>) { // boxed_local lint triggered here
        unimplemented!()
    }       
}

struct SomeObject;

impl ExampleTrait for SomeObject {}

#[allow(unused_variables)]
fn main() {
    let trait_object: Box<dyn ExampleTrait> = Box::new(SomeObject {});
    trait_object.method();
}
```
</details>

<details>
<summary>clippy output</summary>

```
warning: local variable doesn't need to be boxed here
 --> src/main.rs:2:15
  |
2 |     fn method(self: Box<Self>) {
  |               ^^^^
  |
  = note: #[warn(clippy::boxed_local)] on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#boxed_local
```

</details>

**Comment by mikerite** [comment]
`method(self: Box<Self>)` can't be replaced with `method(self)` because the `Self` isn't sized, but couldn't you change it to `method(&self)` or `method(&mut self)` instead?

**Comment by dmarcuse** [comment]
Not for this particular case. The method spawns a thread which takes ownership of `self`, so using a reference wouldn't work as the thread's reference would outlive the object.

**Comment by mikerite** [comment]
The lint is supposed to check if the box is moved inside the body and not warn in that case. In the example below `method2` will get a warning but `method1` won't. 

```
#![warn(clippy::boxed_local)]

use std::mem;

trait ExampleTrait {
    fn method1(self: Box<Self>) { // boxed_local lint is _not_ triggered here
        mem::drop(self)
    }       
    fn method2(self: Box<Self>) { // boxed_local lint triggered here
        unimplemented!();
    }       
}

fn main() {}
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d76b365fa112a716306db92ba56d3eba)

Can you post more of your code?

**Comment by dmarcuse** [comment]
It seems to not detect the value being moved when it's implicitly moved into a closure (e.g. to spawn a thread)

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=233d5551f51d714cbea691ea912979a2)

**Comment by dmarcuse** [comment]
Thank you!
